crates/orderbook/src/database/fee_policy.rs

```diff
@@ -0,0 +1,104 @@
+use {
+    crate::database::Postgres,
+    anyhow::Context,
+    bigdecimal::{BigDecimal, FromPrimitive},
+    database::{auction::AuctionId, OrderUid},
+    model::fee_policy::{FeePolicy, Quote},
+    number::conversions::big_decimal_to_u256,
+};
+
+#[async_trait::async_trait]
+pub trait FeePolicyRetrieving: Send + Sync {
+    async fn fee_policies(
+        &self,
+        auction_id: AuctionId,
+        order_uid: OrderUid,
+    ) -> anyhow::Result<Vec<FeePolicy>>;
+}
+
+#[async_trait::async_trait]
+impl FeePolicyRetrieving for Postgres {
+    async fn fee_policies(
+        &self,
+        auction_id: AuctionId,
+        order_uid: OrderUid,
+    ) -> anyhow::Result<Vec<FeePolicy>> {
+        let mut ex = self.pool.acquire().await?;
+
+        let timer = super::Metrics::get()
+            .database_queries
+            .with_label_values(&["fee_policies"])
+            .start_timer();
+
+        let fee_policies = database::fee_policies::fetch(&mut ex, auction_id, order_uid).await?;
+        timer.stop_and_record();
+
+        let quote = fee_policies
+            .iter()
+            .any(|fp| {
+                matches!(
+                    fp.kind,
+                    database::fee_policies::FeePolicyKind::PriceImprovement
+                )
+            })
+            .then_some({
+                let timer = super::Metrics::get()
+                    .database_queries
+                    .with_label_values(&["order_quote"])
+                    .start_timer();
+
+                let order_quote = database::orders::read_quote(&mut ex, &order_uid)
```

Comment:
These should be collected and then fetched in a single roundtrip imo (can use existing `read_quotes`).

Reply:
Makes sense, reworked.